# Mini Project 5-1 Explore Descriptive Statistics with Python

## **Introduction**

Data professionals often use descriptive statistics to understand the data they are working with and provide collaborators with a summary of the relative location of values in the data, as well an information about its spread. 

For this activity, you are a member of an analytics team for the United States Environmental Protection Agency (EPA). You are assigned to analyze data on air quality with respect to carbon monoxide, a major air pollutant. The data includes information from more than 200 sites, identified by state, county, city, and local site names. You will use Python functions to gather statistics about air quality, then share insights with stakeholders.

## **Step 1: Imports** 


Import the relevant Python libraries `pandas` and `numpy`.

In [4]:
# Import relevant Python libraries.

import pandas as pd
import numpy as np


The dataset provided is in the form of a .csv file named `c4_epa_air_quality.csv`. It contains a susbet of data from the U.S. EPA. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this Project. Please continue with this activity by completing the following instructions.

In [5]:
# RUN THIS CELL TO IMPORT YOUR DATA.

### YOUR CODE HERE
# Display the first five rows of the dataset
df = pd.read_csv("c4_epa_air_quality.csv")
df.head()

Unnamed: 0  date_local    state_name   county_name      city_name  \
0           0  2018-01-01       Arizona      Maricopa        Buckeye   
1           1  2018-01-01          Ohio       Belmont      Shadyside   
2           2  2018-01-01       Wyoming         Teton  Not in a city   
3           3  2018-01-01  Pennsylvania  Philadelphia   Philadelphia   
4           4  2018-01-01          Iowa          Polk     Des Moines   

                                     local_site_name   parameter_name  \
0                                            BUCKEYE  Carbon monoxide   
1                                          Shadyside  Carbon monoxide   
2  Yellowstone National Park - Old Faithful Snow ...  Carbon monoxide   
3                             North East Waste (NEW)  Carbon monoxide   
4                                          CARPENTER  Carbon monoxide   

    units_of_measure  arithmetic_mean  aqi  
0  Parts per million         0.473684    7  
1  Parts per million         0.263158    5  
2  Parts per million         0.111111    2  
3  Parts per million         0.300000    3  
4  Parts per million         0.215789    3

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to read in data from a .csv file and load it into a DataFrame. 

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `read_csv` function from the pandas `library`. The `index_col` parameter can be set to `0` to read in the first column as an index (and to avoid `"Unnamed: 0"` appearing as a column in the resulting DataFrame).

</details>

## **Step 2: Data exploration** 

To understand how the dataset is structured, display the first 10 rows of the data.

In [6]:
# Display first 10 rows of the data.

### YOUR CODE HERE

df.head(10)

Unnamed: 0  date_local    state_name   county_name      city_name  \
0           0  2018-01-01       Arizona      Maricopa        Buckeye   
1           1  2018-01-01          Ohio       Belmont      Shadyside   
2           2  2018-01-01       Wyoming         Teton  Not in a city   
3           3  2018-01-01  Pennsylvania  Philadelphia   Philadelphia   
4           4  2018-01-01          Iowa          Polk     Des Moines   
5           5  2018-01-01        Hawaii      Honolulu  Not in a city   
6           6  2018-01-01        Hawaii      Honolulu  Not in a city   
7           7  2018-01-01  Pennsylvania          Erie           Erie   
8           8  2018-01-01        Hawaii      Honolulu       Honolulu   
9           9  2018-01-01      Colorado       Larimer   Fort Collins   

                                     local_site_name   parameter_name  \
0                                            BUCKEYE  Carbon monoxide   
1                                          Shadyside  Carbon monoxide   
2  Yellowstone National Park - Old Faithful Snow ...  Carbon monoxide   
3                             North East Waste (NEW)  Carbon monoxide   
4                                          CARPENTER  Carbon monoxide   
5                                            Kapolei  Carbon monoxide   
6                                            Kapolei  Carbon monoxide   
7                                                NaN  Carbon monoxide   
8                                           Honolulu  Carbon monoxide   
9                      Fort Collins - CSU - S. Mason  Carbon monoxide   

    units_of_measure  arithmetic_mean  aqi  
0  Parts per million         0.473684    7  
1  Parts per million         0.263158    5  
2  Parts per million         0.111111    2  
3  Parts per million         0.300000    3  
4  Parts per million         0.215789    3  
5  Parts per million         0.994737   14  
6  Parts per million         0.200000    2  
7  Parts per million         0.200000    2  
8  Parts per million         0.400000    5  
9  Parts per million         0.300000    6

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to get a specific number of rows from the top of a DataFrame. 

</details>

**Question:** What does the `aqi` column represent?

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `head()` function from the `pandas` library.

</details>

A: Perhaps Air Quality Index

**Question:** In what units are the aqi values expressed?

A: Parts per Million

Now, get a table that contains some descriptive statistics about the data.

In [8]:
# Get descriptive stats.

### YOUR CODE HERE

df.describe()

Unnamed: 0  arithmetic_mean         aqi
count  260.000000       260.000000  260.000000
mean   129.500000         0.403169    6.757692
std     75.199734         0.317902    7.061707
min      0.000000         0.000000    0.000000
25%     64.750000         0.200000    2.000000
50%    129.500000         0.276315    5.000000
75%    194.250000         0.516009    9.000000
max    259.000000         1.921053   50.000000

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate a table of basic descriptive statistics about the numeric columns in a DataFrame.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `describe()` function from the `pandas` library.

</details>

**Question:** Based on the table of descriptive statistics, what do you notice about the count value for the `aqi` column?

A: That it is 260 which is the same as the other columns and hence there are no missing values

**Question:** What do you notice about the 25th percentile for the `aqi` column?

This is an important measure for understanding where the aqi values lie. 

A: It is 2 which is closer to the 50th percentile than the 75th percentile is to the 50th percentile. Hence a slight rightward skew

**Question:** What do you notice about the 75th percentile for the `aqi` column?

This is another important measure for understanding where the aqi values lie. 

A: It is 9 which is 4 away from the median which is more than than how much the 25th percentile is from the median which is 3. hence a rightward skew

## **Step 3: Statistical tests** 

Next, get some descriptive statistics about the states in the data.

In [10]:
# Get descriptive stats about the states in the data.

### YOUR CODE HERE

df.groupby('state_name').describe()

Unnamed: 0                                                \
                          count        mean         std    min     25%    50%   
state_name                                                                      
Alabama                     1.0   71.000000         NaN   71.0   71.00   71.0   
Alaska                      2.0  240.500000    3.535534  238.0  239.25  240.5   
Arizona                    14.0  131.714286   88.793995    0.0   67.50  140.5   
Arkansas                    1.0  186.000000         NaN  186.0  186.00  186.0   
California                 66.0  137.272727   69.667020   16.0   76.25  144.0   
Colorado                    9.0  119.333333   80.614825    9.0   56.00   97.0   
Connecticut                 4.0  119.250000   94.757146   15.0   82.50  108.5   
Delaware                    1.0  135.000000         NaN  135.0  135.00  135.0   
District Of Columbia        2.0  219.000000   50.911688  183.0  201.00  219.0   
Florida                    12.0  149.000000   66.311935   39.0  109.00  153.0   
Georgia                     2.0  139.000000  125.865007   50.0   94.50  139.0   
Hawaii                      3.0    6.333333    1.527525    5.0    5.50    6.0   
Idaho                       2.0   96.000000   82.024387   38.0   67.00   96.0   
Illinois                    3.0  127.000000   86.156834   28.0   98.00  168.0   
Indiana                     5.0  135.600000   92.629909   11.0   68.00  164.0   
Iowa                        3.0   28.333333   28.711206    4.0   12.50   21.0   
Kansas                      1.0   29.000000         NaN   29.0   29.00   29.0   
Kentucky                    3.0  148.000000   94.873600   84.0   93.50  103.0   
Louisiana                   1.0  152.000000         NaN  152.0  152.00  152.0   
Maine                       2.0  108.500000  101.116270   37.0   72.75  108.5   
Maryland                    4.0  105.000000   82.732501   30.0   66.00   83.5   
Massachusetts               4.0   94.000000   58.109093   36.0   48.00   97.5   
Michigan                    9.0  172.666667   64.625846   65.0  123.00  192.0   
Minnesota                   7.0   94.142857   82.197787   10.0   33.00   73.0   
Mississippi                 1.0  160.000000         NaN  160.0  160.00  160.0   
Missouri                    3.0   72.333333   59.534304   14.0   42.00   70.0   
Montana                     2.0  121.500000    9.192388  115.0  118.25  121.5   
Nebraska                    2.0  162.000000  135.764502   66.0  114.00  162.0   
Nevada                      4.0   55.500000   40.796242   17.0   23.00   52.5   
New Hampshire               2.0  163.500000   24.748737  146.0  154.75  163.5   
New Jersey                  6.0  106.666667   55.913028   32.0   64.75  113.0   
New Mexico                  2.0  199.500000   53.033009  162.0  180.75  199.5   
New York                   10.0  165.800000   43.898367   90.0  134.75  177.5   
North Carolina              4.0  127.000000   92.372435   54.0   69.00   97.5   
North Dakota                1.0  108.000000         NaN  108.0  108.00  108.0   
Ohio                       12.0  128.666667   96.437765    1.0   43.75  134.5   
Oklahoma                    4.0  131.750000   49.473057   85.0   91.00  133.5   
Oregon                      2.0  215.500000   53.033009  178.0  196.75  215.5   
Pennsylvania               10.0  103.200000   75.271066    3.0   67.75   96.5   
Puerto Rico                 1.0  187.000000         NaN  187.0  187.00  187.0   
Rhode Island                2.0   83.500000   85.559921   23.0   53.25   83.5   
South Carolina              1.0  221.000000         NaN  221.0  221.00  221.0   
South Dakota                1.0   53.000000         NaN   53.0   53.00   53.0   
Tennessee                   2.0  127.000000  110.308658   49.0   88.00  127.0   
Texas                      10.0  138.900000   71.242076   24.0   93.75  125.5   
Utah                        3.0  171.666667   50.332230  125.0  145.00  165.0   
Vermont                     3.0  100.666667   55.048

<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `pandas` library that allows you to generate basic descriptive statistics about a DataFrame or a column you are interested in.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

 Use the `describe()` function from the `pandas` library. Note that this function can be used:
- "on a DataFrame (to find descriptive statistics about the numeric columns)" 
- "directly on a column containing categorical data (to find pertinent descriptive statistics)"

</details>

**Question:** What do you notice while reviewing the descriptive statistics about the states in the data? 

Note: Sometimes you have to individually calculate statistics. To review to that approach, use the `numpy` library to calculate each of the main statistics in the preceding table for the `aqi` column.

A: Some states have missing std values. Also the mean and median of different states are very away from other states indicating incosistent aqi ratings

## **Step 4. Results and evaluation**

Now, compute the mean value from the `aqi` column.

In [14]:
# Compute the mean value from the aqi column.

### YOUR CODE HERE

mean_aqi = df['aqi'].mean()

print("Mean AQI:", mean_aqi)

Mean AQI: 6.757692307692308


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the mean value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `mean()` function from the `numpy` library.

</details>

**Question:** What do you notice about the mean value from the `aqi` column?

This is an important measure, as it tells you what the average air quality is based on the data.

A: It is quite low compared to the range in the values

Next, compute the median value from the aqi column.

In [15]:
# Compute the median value from the aqi column.

### YOUR CODE HERE

median_aqi = df['aqi'].median()

print("Median AQI:", median_aqi)

Median AQI: 5.0


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the median value from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `median()` function from the `numpy` library.

</details>

**Question:** What do you notice about the median value from the `aqi` column?

This is an important measure for understanding the central location of the data.

A: Slightly lower than the mean implying there are some outlier values to the right

Next, identify the minimum value from the `aqi` column.

In [16]:
# Identify the minimum value from the aqi column.

### YOUR CODE HERE

min_aqi = df['aqi'].min()

print("Min AQI:", min_aqi)

Min AQI: 0


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the minimum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `min()` function from the `numpy` library.

</details>

**Question:** What do you notice about the minimum value from the `aqi` column?

This is an important measure, as it tell you the best air quality observed in the data.

The minimum value for the aqi column is 0. This means that the smallest aqi value in the data is 0 parts per million.


Now, identify the maximum value from the `aqi` column.

In [17]:
# Identify the maximum value from the aqi column.

### YOUR CODE HERE
max_aqi = df['aqi'].max()

print("Max AQI:", max_aqi)

Max AQI: 50


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the maximum value from an array or a Series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

  Use the `max()` function from the `numpy` library.

</details>

**Question:** What do you notice about the maximum value from the `aqi` column?

This is an important measure, as it tells you which value in the data corresponds to the worst air quality observed in the data.

A: This is 50 which means that some outliers exist and there are very far from the mean

Now, compute the standard deviation for the `aqi` column.

By default, the `numpy` library uses 0 as the Delta Degrees of Freedom, while `pandas` library uses 1. To get the same value for standard deviation using either library, specify the `ddof` parameter to 1 when calculating standard deviation.

In [18]:
# Compute the standard deviation for the aqi column.

### YOUR CODE HERE

std_aqi = df['aqi'].std()

print("Std AQI:", std_aqi)

Std AQI: 7.061706678820724


<details>
  <summary><h4><strong>Hint 1</strong></h4></summary>

  There is a function in the `numpy` library that allows you to get the standard deviation from an array or a series of values.

</details>

<details>
  <summary><h4><strong>Hint 2</strong></h4></summary>

Use the `std()` function from the `numpy` library. Make sure to specify the `ddof` parameter as 1. To read more about this function,  refer to its documentation in the references section of this lab.

</details>

**Question:** What do you notice about the standard deviation for the `aqi` column? 

This is an important measure of how spread out the aqi values are.

A: Fairly high standard deviation which means they are somewhat spread out

## **Considerations**


**What are some key takeaways that you learned during this Project?**

A: That it is fairly easy to use different functions to analyze a dataset

**How would you present your findings from this Project to others? Consider the following relevant points noted by AirNow.gov as you respond:**
- "AQI values at or below 100 are generally thought of as satisfactory. When AQI values are above 100, air quality is considered to be unhealthy—at first for certain sensitive groups of people, then for everyone as AQI values increase."
- "An AQI of 100 for carbon monoxide corresponds to a level of 9 parts per million."

A: I would list the AQis of different staes and then talk about the aggregate descriptor AQI of the entire data set. Mention min, 25th percentile, median, and 75th percentile, and max, and mention the units being parts per million.

**What summary would you provide to readers? Use the same information provided previously from AirNow.gov as you respond.**

A: The Air Quality Index (AQI) data shoes a range of air quality levels across different states. The lowest recorded AQI value is 0 while the highest reaches 50. The Median is 5. According to AirNOw.gov, AQI values above 100 indicate unhealthy air quality, particular for sensitive groups at first. If any states have an AQI above 100, there might be health risks associated

Additionally, an AQI of 100 for carbon monoxide corresponds to 9 parts per million.

**References**

[Air Quality Index - A Guide to Air Quality and Your Health](https://www.airnow.gov/sites/default/files/2018-04/aqi_brochure_02_14_0.pdf). (2014,February)

[Numpy.Std — NumPy v1.23 Manual](https://numpy.org/doc/stable/reference/generated/numpy.std.html)

US EPA, OAR. (2014, 8 July).[*Air Data: Air Quality Data Collected at Outdoor Monitors Across the US*](https://www.epa.gov/outdoor-air-quality-data). 